Swagger changes for MFA work

deymadhumanti opened the pull request:
<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow. 
</i>

### Changelog
Add a changelog entry for this PR by answering the following questions:
  1. What's the purpose of the update?
      - [ ] new service onboarding
      - [x] new API version
      - [ ] update existing version for new feature
      - [ ] update existing version to fix swagger quality issue in s360
      - [ ] Other, please clarify
  2. When are you targeting to deploy the new service/feature to public regions? Please provide the date or, if the date is not yet available, the month.
  
We are planning to deploy the feature to all public regions by Mid of August (Tentative Date is 23rd August, 2021 ). Swagger and ARM manifest changes needs to be published before this.

  3. When do you expect to publish the swagger? Please provide date or, the the date is not yet available, the month.
  By mid of August. (Tentative Date is 23rd August, 2021 )
  4. If updating an existing version, please select the specific langauge SDKs and CLIs that must be refreshed after the swagger is published.
      - [ ] SDK of .NET (need service team to ensure code readiness)
      - [ ] SDK of Python
      - [ ] SDK of Java
      - [ ] SDK of Js
      - [ ] SDK of Go
      - [ ] PowerShell
      - [ ] CLI
      - [ ] Terraform
      - [ ] No refresh required for updates in this PR

### Contribution checklist:
- [x] I commit to follow the [Breaking Change Policy](http://aka.ms/bcforapi) of "no breaking changes"
- [x] I have reviewed the [documentation](https://aka.ms/ameonboard) for the workflow.
- [x] [Validation tools](https://aka.ms/swaggertools) were run on swagger spec(s) and errors have all been fixed in this PR. [How to fix?](https://aka.ms/ci-fix)

If any further question about AME onboarding or validation tools, please view the [FAQ](https://aka.ms/faqinprreview).

### ARM API Review Checklist 

> **Applicability**: :warning: 
>
> If your changes encompass only the following scenarios, you should SKIP this section, as these scenarios do not require ARM review.
> - Change to data plane APIs 
> - Adding new properties 
> - All removals

Otherwise your PR may be subject to ARM review requirements. Complete the following:
- [x] Check this box if any of the following apply to the PR so that label “WaitForARMFeedback” will be added automatically to begin ARM API Review. Failure to comply may result in delays to the manifest.
  - Adding a new service
  - Adding new API(s)
  - Adding a new API version
    -[ ] To review changes efficiently, ensure you copy the existing version into the new directory structure for first commit (including refactoring) and then push new changes, including version updates, in separate commits.

- [x] Ensure you've reviewed following [guidelines](https://aka.ms/rpguidelines) including [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md). Estimated time (4 hours). This is required before you can request review from ARM API Review board.

- [ ] If you are blocked on ARM review and want to get the PR merged with urgency, please get the ARM oncall for reviews (*RP Manifest Approvers* team under <ins>Azure Resource Manager service</ins>) from IcM and reach out to them. 

### Breaking Change Review Checklist 
If any of the following scenarios apply to the PR, request approval from the Breaking Change Review Board as defined in the [Breaking Change Policy](http://aka.ms/bcforapi). 
- [ ] Removing API(s) in a stable version
- [ ] Removing properties in a stable version
- [ ] Removing API version(s) in a stable version
- [ ] Updating API in a stable or public preview version with Breaking Change Validation errors
- [ ] Updating API(s) in public preview over 1 year (refer to [Retirement of Previews](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37683/Retirement-of-Previews))

**Action**: to initiate an evaluation of the breaking change, create a new intake using the [template for breaking changes](https://aka.ms/Breakingchangetemplate). Addition details on the process and office hours are on the [Breaking change Wiki](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37684/Breaking-Changes).

Please follow the link to find more details on [PR review process](https://aka.ms/SwaggerPRReview).


**Description of the PR Changes :**


Merged the existing swagger json files (2021-06-01 for recoveryservicesbackup and 2021-01-01 for dataprotection) to create the base swagger json files with API version 2021-07-01 in first commit.

In the second commit, made the API version update in Swagger and example json files, readme.md file and the new changes.

We have created below work item for approval of breaking changes.

**Scenario 10554276**: RecoveryServicesBackup-BreakingChangeApproval

**Created a suppression request with below details for Lintdiff failures.**

![image](https://user-images.githubusercontent.com/58032062/128731613-3bb53a46-cce2-4eca-ad28-5b75d3731aa4.png)

**All the new changes in this PR are already reviewed and checked-in in the preview folder.
For dataprotection and recoveryservicesbackup, preview version was 2021-02-01-preview. Changes from this preview version folder are merged in the second commit of this PR.**

madhude committed:
Baseline changes for API version 2021-07-01

madhude committed:
Made the new changes related to API version 2021-07-01

madhude committed:
Fixing Lintdiff failure

mentat9 commented:
>           "202": {

Operation results API should always be synchronous. Did you intend to add this to the responses from the backupVaults path?

---
Refers to: specification/dataprotection/resource-manager/Microsoft.DataProtection/stable/2021-07-01/dataprotection.json:480 in 38224a9. [](commit_id = 38224a9e8e5a3a9b647b3a3e192b31ccd0d019d0, deletion_comment = False)

mentat9 commented:
>         "produces": [

Don't need this here or on other paths: it's already covered by the higher-level declaration.

---
Refers to: specification/dataprotection/resource-manager/Microsoft.DataProtection/stable/2021-07-01/dataprotection.json:1909 in 38224a9. [](commit_id = 38224a9e8e5a3a9b647b3a3e192b31ccd0d019d0, deletion_comment = False)

mentat9 commented:
>         "consumes": [

Not needed here.

---
Refers to: specification/dataprotection/resource-manager/Microsoft.DataProtection/stable/2021-07-01/dataprotection.json:1996 in 38224a9. [](commit_id = 38224a9e8e5a3a9b647b3a3e192b31ccd0d019d0, deletion_comment = False)

mentat9 commented:
>     "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.DataProtection/resourceGuards/{resourceGuardsName}/{operationName}": {

This path doesn't follow ARM resource naming rules. You need a resource type to identify instances of the operations, e.g. .../{resourceGuardsName}/operations/{operationName}.

---
Refers to: specification/dataprotection/resource-manager/Microsoft.DataProtection/stable/2021-07-01/dataprotection.json:2200 in 38224a9. [](commit_id = 38224a9e8e5a3a9b647b3a3e192b31ccd0d019d0, deletion_comment = False)

mentat9 commented:
>     "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.DataProtection/resourceGuards/{resourceGuardsName}/{operationName}/{operationRequestName}": {

Doesn't follow ARM naming rules.

---
Refers to: specification/dataprotection/resource-manager/Microsoft.DataProtection/stable/2021-07-01/dataprotection.json:2259 in 38224a9. [](commit_id = 38224a9e8e5a3a9b647b3a3e192b31ccd0d019d0, deletion_comment = False)

mentat9 commented:
>         "lastLockUpdateTime": {

Is this the same as LastModifiedAt in systemData? If so, then this should be removed.

---
Refers to: specification/dataprotection/resource-manager/Microsoft.DataProtection/stable/2021-07-01/dataprotection.json:5151 in 38224a9. [](commit_id = 38224a9e8e5a3a9b647b3a3e192b31ccd0d019d0, deletion_comment = False)

mentat9 commented:
>     "ObjectResource": {

Non-schematized/empty objects are not recommended. They are hard for customers to do anything with, and can't be usd with many ARM features like ARG and Policy.

---
Refers to: specification/dataprotection/resource-manager/Microsoft.DataProtection/stable/2021-07-01/dataprotection.json:5251 in 38224a9. [](commit_id = 38224a9e8e5a3a9b647b3a3e192b31ccd0d019d0, deletion_comment = False)

deymadhumanti commented:
> > ```
> >     "lastLockUpdateTime": {
> > ```
> 
> Is this the same as LastModifiedAt in systemData? If so, then this should be removed.
> 
> Refers to: specification/dataprotection/resource-manager/Microsoft.DataProtection/stable/2021-07-01/dataprotection.json:5151 in [38224a9](https://github.com/Azure/azure-rest-api-specs/commit/38224a9e8e5a3a9b647b3a3e192b31ccd0d019d0). [](commit_id = [38224a9](https://github.com/Azure/azure-rest-api-specs/commit/38224a9e8e5a3a9b647b3a3e192b31ccd0d019d0), deletion_comment = False)

Removed this. here lastLockUpdateTime indicates the last time when resource guard resource was updated

madhude committed:
resolved PR comments

madhude committed:
Fixing validation failures

madhude committed:
Fix the validation failures

madhude committed:
Made lintdiff failures

mentat9 commented:
>           "202": {

Resolving per RPC link: [https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/async-api-reference.md#202-accepted-and-location-headers](https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/async-api-reference.md#202-accepted-and-location-headers)

---
In reply to: [896285342](https://github.com/Azure/azure-rest-api-specs/pull/15577#issuecomment-896285342) [](http://example.com/codeflow?ancestors=896285342,896285342)

---
Refers to: specification/dataprotection/resource-manager/Microsoft.DataProtection/stable/2021-07-01/dataprotection.json:480 in 38224a9. [](commit_id = 38224a9e8e5a3a9b647b3a3e192b31ccd0d019d0, deletion_comment = False)

mentat9 commented:
>     "ObjectResource": {

Please don't publish placeholder objects in your API for future use. To complete development of the API you will need a new API version anyway, so the empty objects aren't doing anything except confusing consumers of your current API.

---
In reply to: [896286506](https://github.com/Azure/azure-rest-api-specs/pull/15577#issuecomment-896286506) [](http://example.com/codeflow?ancestors=896286506,896286506)

---
Refers to: specification/dataprotection/resource-manager/Microsoft.DataProtection/stable/2021-07-01/dataprotection.json:5251 in 38224a9. [](commit_id = 38224a9e8e5a3a9b647b3a3e192b31ccd0d019d0, deletion_comment = False)

mentat9 commented:
>     "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.DataProtection/resourceGuards/{resourceGuardsName}/{operationName}": {

This is not a good approach from ARM perspective. This kind of path is half resource and half resource type. ARM features like Azure Policy and ARG will not be able to deal with these. I think the better ARM-friendly approach would be to use a $expand parameter for this.

---
In reply to: [896286087](https://github.com/Azure/azure-rest-api-specs/pull/15577#issuecomment-896286087) [](http://example.com/codeflow?ancestors=896286087,896286087)

---
Refers to: specification/dataprotection/resource-manager/Microsoft.DataProtection/stable/2021-07-01/dataprotection.json:2200 in 38224a9. [](commit_id = 38224a9e8e5a3a9b647b3a3e192b31ccd0d019d0, deletion_comment = False)

mentat9 commented:
>     "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.DataProtection/resourceGuards/{resourceGuardsName}/{operationName}/{operationRequestName}": {

Same comment as above. It's not really just naming: it's about fitting the ARM resource model properly and being consistent with the rest of Azure APIs.

---
In reply to: [896286201](https://github.com/Azure/azure-rest-api-specs/pull/15577#issuecomment-896286201) [](http://example.com/codeflow?ancestors=896286201,896286201)

---
Refers to: specification/dataprotection/resource-manager/Microsoft.DataProtection/stable/2021-07-01/dataprotection.json:2259 in 38224a9. [](commit_id = 38224a9e8e5a3a9b647b3a3e192b31ccd0d019d0, deletion_comment = False)

deymadhumanti commented:
> > ```
> >       "202": {
> > ```
> 
> Operation results API should always be synchronous. Did you intend to add this to the responses from the backupVaults path?
> 
> In reply to: [896285342](https://github.com/Azure/azure-rest-api-specs/pull/15577#issuecomment-896285342)
> 
> Refers to: specification/dataprotection/resource-manager/Microsoft.DataProtection/stable/2021-07-01/dataprotection.json:480 in [38224a9](https://github.com/Azure/azure-rest-api-specs/commit/38224a9e8e5a3a9b647b3a3e192b31ccd0d019d0). [](commit_id = [38224a9](https://github.com/Azure/azure-rest-api-specs/commit/38224a9e8e5a3a9b647b3a3e192b31ccd0d019d0), deletion_comment = False)

These changes are intended. As per this ARM API guideline wiki (https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/async-api-reference.md#202-accepted-and-location-headers), below thing is mentioned in point 4.
"If the operation is not complete, return 202 (Accepted) again with Location header and optionally an updated Retry-After header."

deymadhumanti commented:
> > ```
> >     "consumes": [
> > ```
> 
> Not needed here.
> 
> In reply to: [896285634](https://github.com/Azure/azure-rest-api-specs/pull/15577#issuecomment-896285634)
> 
> Refers to: specification/dataprotection/resource-manager/Microsoft.DataProtection/stable/2021-07-01/dataprotection.json:1996 in [38224a9](https://github.com/Azure/azure-rest-api-specs/commit/38224a9e8e5a3a9b647b3a3e192b31ccd0d019d0). [](commit_id = [38224a9](https://github.com/Azure/azure-rest-api-specs/commit/38224a9e8e5a3a9b647b3a3e192b31ccd0d019d0), deletion_comment = False)

Actually we are using some tool to generate swagger json from service code. These sections are getting added automatically for each API.  These changes are already there in the existing json files for recoveryservicesbackup and dataprotection RP. We need to update the tool to cleanup these. Otherwise, manual cleanup is needed which is little bit error-prone.  We will work on fixing this tool. Can you please approve these changes for this PR.

deymadhumanti commented:
> > ```
> >     "produces": [
> > ```
> 
> Don't need this here or on other paths: it's already covered by the higher-level declaration.
> 
> In reply to: [896285534](https://github.com/Azure/azure-rest-api-specs/pull/15577#issuecomment-896285534)
> 
> Refers to: specification/dataprotection/resource-manager/Microsoft.DataProtection/stable/2021-07-01/dataprotection.json:1909 in [38224a9](https://github.com/Azure/azure-rest-api-specs/commit/38224a9e8e5a3a9b647b3a3e192b31ccd0d019d0). [](commit_id = [38224a9](https://github.com/Azure/azure-rest-api-specs/commit/38224a9e8e5a3a9b647b3a3e192b31ccd0d019d0), deletion_comment = False)

Actually we are using some tool to generate swagger json from service code. These sections are getting added automatically for each API.  These changes are already there in the existing json files for recoveryservicesbackup and dataprotection RP. We need to update the tool to cleanup these. Otherwise, manual cleanup is needed which is little bit error-prone.  We will work on fixing this tool. Can you please approve these changes for this PR.

madhude committed:
Resolved PR comments

madhude committed:
Fixing validation failures

deymadhumanti commented:
> > ```
> > "ObjectResource": {
> > ```
> 
> Please don't publish placeholder objects in your API for future use. To complete development of the API you will need a new API version anyway, so the empty objects aren't doing anything except confusing consumers of your current API.
> 
> In reply to: [896286506](https://github.com/Azure/azure-rest-api-specs/pull/15577#issuecomment-896286506)
> 
> Refers to: specification/dataprotection/resource-manager/Microsoft.DataProtection/stable/2021-07-01/dataprotection.json:5251 in [38224a9](https://github.com/Azure/azure-rest-api-specs/commit/38224a9e8e5a3a9b647b3a3e192b31ccd0d019d0). [](commit_id = [38224a9](https://github.com/Azure/azure-rest-api-specs/commit/38224a9e8e5a3a9b647b3a3e192b31ccd0d019d0), deletion_comment = False)

Updated the json file with the fix. Removed the placeholder empty object

deymadhumanti commented:
> > ```
> > "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.DataProtection/resourceGuards/{resourceGuardsName}/{operationName}/{operationRequestName}": {
> > ```
> 
> Same comment as above. It's not really just naming: it's about fitting the ARM resource model properly and being consistent with the rest of Azure APIs.
> 
> In reply to: [896286201](https://github.com/Azure/azure-rest-api-specs/pull/15577#issuecomment-896286201)
> 
> Refers to: specification/dataprotection/resource-manager/Microsoft.DataProtection/stable/2021-07-01/dataprotection.json:2259 in [38224a9](https://github.com/Azure/azure-rest-api-specs/commit/38224a9e8e5a3a9b647b3a3e192b31ccd0d019d0). [](commit_id = [38224a9](https://github.com/Azure/azure-rest-api-specs/commit/38224a9e8e5a3a9b647b3a3e192b31ccd0d019d0), deletion_comment = False)

Fixed this.

deymadhumanti commented:
> > ```
> > "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.DataProtection/resourceGuards/{resourceGuardsName}/{operationName}": {
> > ```
> 
> This is not a good approach from ARM perspective. This kind of path is half resource and half resource type. ARM features like Azure Policy and ARG will not be able to deal with these. I think the better ARM-friendly approach would be to use a $expand parameter for this.
> 
> In reply to: [896286087](https://github.com/Azure/azure-rest-api-specs/pull/15577#issuecomment-896286087)
> 
> Refers to: specification/dataprotection/resource-manager/Microsoft.DataProtection/stable/2021-07-01/dataprotection.json:2200 in [38224a9](https://github.com/Azure/azure-rest-api-specs/commit/38224a9e8e5a3a9b647b3a3e192b31ccd0d019d0). [](commit_id = [38224a9](https://github.com/Azure/azure-rest-api-specs/commit/38224a9e8e5a3a9b647b3a3e192b31ccd0d019d0), deletion_comment = False)

Fixed this in the latest PR iteration

mentat9 commented:
Please work with this week's ARM API review oncall (@pilor) to finish the review.

pilor commented on specification/dataprotection/resource-manager/Microsoft.DataProtection/stable/2021-07-01/examples/ResourceGuardCRUD/GetDefaultOperationsRequestObject.json:
```diff
@@ -0,0 +1,18 @@
+      "body": {
+        "id": "subscriotions/0b352192-dcac-4cc7-992e-a96190ccc68c/resourceGroups/SampleResourceGroup/providers/Microsoft.DataProtection/resourceGuards/swaggerExample/disableSoftDeleteRequests/default",
+        "name": "default",
+        "type": "Microsoft.DataProtection/resourceGuards/deleteResourceGuardProxyRequests"
```
This is not the right type for the resource ID.  Is this also a real example of the response from these GETs?  What value are they providing?

deymadhumanti replied:
> This is not the right type for the resource ID. Is this also a real example of the response from these GETs? What value are they providing?

Hi, This Type, we are getting from GET API call response. These changes are intended. Here type indicated the type of the critical operation request is protected by the ResourceGuard.

Here, Microsoft.DataProtection is provider namespace. resourceGuards is the type of resource used to protect the critical operation. deleteResourceGuardProxyRequests is critical operation request. Here, deletion of ResourceGuardProxy is the critical operation.

pilor replied:
The "type" in your example is "Microsoft.DataProtection/resourceGuards/deleteResourceGuardProxyRequests" but the resource ID is of type "Microsoft.DataProtection/resourceGuards/disableSoftDeleteRequests".  Please go through all the examples and make sure they are correct.  

What does this resource tell the user?  There are no resource specific properties.  Is the existence of one of these resources what triggers the resource guard to "guard" something?  If so, why doesn't the resourceGuard itself just have properties that enable/disable these features?

deymadhumanti replied:
Fixed and rechecked all the example json files. 

Here Backup Admin owns the vault. And that Vault will be protected by ResourceGuard resource. ResourceGuard resource will be created under security admin. Thus when any critical operation will be performed in the vault, permission is needed from both the admin, Thus we are enforcing multi user authentication with this feature. Here one to one mapping is there between a resource (vault) and resourceGuard. This request object is needed for requestor approval  model. Details are mentioned https://microsoft.sharepoint.com/:w:/t/AzureBackupTeam/EXw8RJ8P9KRNr5f7uEwE3swB-YfMbddSrwp6J7-FRVaudA?e=pbr0kQ&CID=086C017D-8298-43B9-97E7-BFCAC023F5A3&wdLOR=cDC5E2EEE-AC6A-4E35-A61B-08A7E05E2BCC

madhude committed:
Resolved PR comments

BigCat20196 commented:
/azp run
